run_train.py

```diff
@@ -117,11 +201,28 @@ def get_dataloader(trainer: DistributedTrainer):
                 f"Try train_steps<={len(dataloader.dataset) // trainer.global_batch_size + trainer.start_iteration_step}"
             )
     else:
-        raise ValueError(f"Unhandled case of `self.config.data.dataset`. Got: {trainer.config.data.dataset}")
+        raise ValueError(f"Unhandled case of `self.config.data.dataset`. Got: {dataset}")
 
     return dataloader
 
 
+def get_dataloader_for_training_stages(trainer: DistributedTrainer) -> Dict[str, DataLoader]:
+    trainer.config.data = cast(DataArgs, trainer.config.data)
+    if trainer.config.data.dataset_stages is None:
+        return {"default": get_dataloader(trainer, trainer.config.data.dataset)}
+    else:
+        sorted_stages = sorted(trainer.config.data.dataset_stages, key=lambda stage: stage.training_steps)
+        dataloaders = {}
+        for idx, stage in enumerate(sorted_stages):
+            dataloader = (
+                get_dataloader(trainer, stage.dataset)
+                if idx == 0
+                else lambda stage=stage: get_dataloader(trainer, stage.dataset)
+            )
+            dataloaders[stage.name] = dataloader
+        return dataloaders
```

Comment:
How does this compare with having a single dataloader (so no need to change our trainer dataloader logic)?

Reply:
- CPU/Memory: does allocating n dataloaders mean we allocate n times the number of workers?
- Determinism: Are we sure we keep determinism when reloading? Let’s say we remove the 1st training stage, are we sure the dataloader will still feed the same data?

Reply:
> Are we sure we keep determinism when reloading? Let’s say we remove the 1st training stage, are we sure the dataloader will still feed the same data?
the second dataloader use a different dataset, it use the same seed in the data section, so as long as the another run use the same seed, it should the same (same for the first dataloader)

Reply:
> does allocating n dataloaders mean we allocate n times the number of workers?
will add remove the earlier dataloaders... and destroy its workers...